### ISSUE TITLE
Errors that happen in onProgress result in repeated HEAD requests

### ISSUE THREAD

**Issue by tdewolff** [tus/tus-js-client#696]
If there is an error in the onProgress callback, instead of reporting the error in the console, the JS client sends out HEAD requests (instead of PATCH requests) which is quite confusing!

Example:
```js
      let upload = new tus.Upload(file, {
        endpoint: '/url',
        retryDelays: [0, 3000, 5000, 10000, 20000],
         metadata: {
          name: file.name,
          type: file.type,
          modtime: file.lastModified,
        },
        onError: function (error) {
          console.log(error);
        },
        onProgress: function (bytesUploaded, bytesTotal) {
          (null).method; // ERROR
        },
        onSuccess: function (e) {
          console.log('Download %s from %s', upload.file.name, upload.url)
        },
      });
```

**Comment by Acconut** [comment]
That's right, good point. We don't have errors that are thrown inside user callbacks well. This is likely not limited to `onProgress`, but is also the case for `onSuccess`, `onError`, and `onShouldRetry`. I wonder what the best way to handle those errors is:
- errors caught from user callbacks could trigger the `onError`, signaling that the upload was stopped because of this error. If no `onError` callback is registered, the errors is directly thrown.
- errors caught from user callbacks are logged to the console. While this allows the upload to continue, such errors could easily be missed, especially when an application produces a significant amount of logs, making debugging much harder. In addition, there would be no way for the user to handle those errors.

My first temptation would be to go with the first approach, which makes error handling more present. What do you think?

**Comment by tdewolff** [comment]
Thanks Marius for the quick reply! My initial expectation (adhering to the principle of least-surprise) would be that errors in callbacks are simply not caught/handled (and thus appear in the console as an error by default). Perhaps this would mean re-throwing the error. Otherwise, option one seems good!

**Comment by h-sigma** [comment]
Thoughts on suggested methods: 
1. How would you handle errors thrown from inside `onError`? Also, this way, you have different behavior based on whether onError is defined or not, which needs you to go read the manual and be sure of how it works.
2. It is highly likely that errors from inside of these callbacks are due to programmer error, and not because of a network/library issue. 

Based on these, I would suggest that the library should re-throw errors from user callbacks, but the library itself should ignore the errors, i.e. the errors should not cause it to go into retry mode or anything like that. If the user wants to fail an upload, they should be doing it explicitly via either calling a method on the upload instance, or through a callback passed to the `onChunk` and other user callbacks.

**Comment by ixxie** [comment]
> Based on these, I would suggest that the library should re-throw errors from user callbacks, but the library itself should ignore the errors, i.e. the errors should not cause it to go into retry mode or anything like that. If the user wants to fail an upload, they should be doing it explicitly via either calling a method on the upload instance, or through a callback passed to the `onChunk` and other user callbacks.

As someone who just got very confused by this issue, I totally agree the errors should be rethrown. Its absolutely essential during debugging to be able to see the error you caused.
